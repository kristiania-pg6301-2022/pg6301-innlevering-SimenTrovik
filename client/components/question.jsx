import { useEffect, useState } from "react";
import React from "react";

export function Question({ alterCounter, questionApi }) {
  const [loading, setLoading] = useState(true);
  const [question, setQuestion] = useState({});
  const [chosen, setChosen] = useState("");

  // Vi føler dette er en litt weird måte å få lastet et nytt spørsmål etter brukeren har svart
  // Du må gjerne gi tilbakemelding på en bedre løsning :)
  const [answered, setAnswered] = useState(0);

  useEffect(async () => {
    setLoading(true);
    setQuestion(await questionApi.getQuestion());
    setLoading(false);
  }, [answered]);

  if (loading) {
    return <div>Loading...</div>;
  }

  async function handleSubmit(event) {
    event.preventDefault();
    const result = await questionApi.checkAnswer(question.id, chosen);
    if (result.isCorrect) {
      alterCounter(1);
    } else {
      alterCounter(-1);
    }
    setAnswered(answered + 1);
  }

  return (
    <div>
      <div className={"top"}>{question.question}</div>
      <form onSubmit={handleSubmit}>
        {Object.entries(question.answers).map(([key, val]) => {
          if (val === null) {
            return;
          }
          return (
            <div key={key}>
              <input
                type={"radio"}
                name={"answer"}
                value={key}
                onClick={(e) => {
                  setChosen(key);
                }}
              />
              {val}
            </div>
          );
        })}
        <button>Submit</button>
      </form>
    </div>
  );
}
