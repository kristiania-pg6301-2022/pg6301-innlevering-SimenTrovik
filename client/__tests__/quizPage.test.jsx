import TEST_QUESTION from "./testdata/TEST_QUESTION.json";
import { Question } from "../components/question.jsx";
import ReactDOM from "react-dom";
import React from "react";
import { act } from "react-dom/test-utils";

async function getQuestion() {
  return {
    id: TEST_QUESTION.id,
    category: TEST_QUESTION.category,
    question: TEST_QUESTION.question,
    answers: TEST_QUESTION.answers,
  };
}

const defaultQuestionApi = {
  getQuestion,
  checkAnswer: jest.fn(),
};

describe("question", () => {
  it("renders a question", async () => {
    const element = document.createElement("div");
    await act(async () => {
      await ReactDOM.render(
        <Question questionApi={defaultQuestionApi} />,
        element
      );
    });
    expect(element).toMatchSnapshot();
  });
});
