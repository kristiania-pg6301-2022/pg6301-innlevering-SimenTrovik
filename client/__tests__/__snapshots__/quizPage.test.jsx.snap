// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`question renders a question 1`] = `
<div>
  <div>
    <div
      class="top"
    >
      What is the correct JavaScript syntax to change the content of the HTML element below?
    </div>
    <form>
      <div>
        <input
          name="answer"
          type="radio"
          value="answer_a"
        />
        #demo.innerHTML = "Hello World!";
      </div>
      <div>
        <input
          name="answer"
          type="radio"
          value="answer_b"
        />
        document.getElementById("demo").innerHTML = "Hello World!";
      </div>
      <div>
        <input
          name="answer"
          type="radio"
          value="answer_c"
        />
        document.getElement("p").innerHTML = "Hello World!";
      </div>
      <div>
        <input
          name="answer"
          type="radio"
          value="answer_d"
        />
        document.getElementByName("p").innerHTML = "Hello World!";
      </div>
      <button>
        Submit
      </button>
    </form>
  </div>
</div>
`;
